### PR TITLE
Add a star badge for already bookmarked websites

### LINKED ISSUE #30: Graphical feedback of linkding bookmark button
It would be nice if the linkding bookmark button would show some kind of graphical feedback if the website you're on has already been bookmarked.

### PR DESCRIPTION
**Pull request by jonathan-s**
Fixes #30 
This PR extracts the code for the graphical feedback which was added in #46. So the basis for this code was graciously provided by @gantoine.  
Screenshot 2024-01-20 at 11 06 18

### CODE REVIEW

**Review by gantoine — COMMENTED** [review]
👌🏼

File: src/background.js [hunk 1]
```diff
@@ -25,6 +25,49 @@ async function initApi() {
...
+  browser.browserAction.setBadgeText({ text: "", tabId: tabId });
+  browser.browserAction.setTitle({ title: "Save bookmark", tabId: tabId });
+}
+
+async function setDymamicBadge(tabId) {
+  const badgeText = await browser.browserAction.getBadgeText({ tabId });
+  const starred = badgeText === "★";
+
+  const tab = await browser.tabs.get(tabId);
+  const tabMetadata = await loadTabMetadata(tab.url);
```
Comment (sissbruecker) [inline src/background.js:46]: This will fetch the metadata for every single navigated URL from the linkding server, which was disabled by default as part of a previous issue and PR:
- https://github.com/sissbruecker/linkding-extension/issues/36
- https://github.com/sissbruecker/linkding-extension/pull/38

This new feature should only be active if the respective option for pre-caching requests is enabled (`precacheEnabled`). The option can be respected by passing `true` as the second parameter: `loadTabMetadata(tab.url, true)`.

The options page should also be modified somehow so that it becomes apparent that the star badge will only work if the option is enabled.

File: src/background.js [hunk 2]
```diff
@@ -25,6 +25,49 @@ async function initApi() {
   return api !== null;
 }
 
+/* Dynamic badge */
+function setStarredBadge(tabId) {
+  browser.browserAction.setBadgeText({ text: "★", tabId: tabId });
```
Comment (sissbruecker) [inline src/background.js:30]: This API doesn't work in Chrome, where Manifest V3 is used:
```
Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'getBadgeText')
```

I would suggest to move the logic for setting the badge into `browser.js` and provide separate implementations for Chrome and Firefox using either `browser.action` (Chrome) or `browser.browserAction` (Firefox).

File: src/background.js (hunk 1 above)
Comment (sissbruecker) [reply src/background.js:46]: There are also existing listeners for tab activated / updated, which already prefetch the metadata. It would be good if the extension wouldn't fetch the metadata twice when a tab is activated / updated.

File: src/background.js (hunk 2 above)
Comment (jonathan-s) [reply src/background.js:30]: @sissbruecker Regarding this, I wonder why we have a separate branch for chrome with manifest v3, and on master using v2, as this is related to which manifest version you are using. As of [firefox version 109](https://extensionworkshop.com/documentation/develop/manifest-v3-migration-guide/), it supports manifest version v3. Version 109 was released in January 2023.

File: src/background.js [hunk 3]
```diff
@@ -25,6 +25,49 @@ async function initApi() {
...
+  browser.browserAction.setBadgeTextColor({ color: "#FFE234", tabId: tabId });
+  browser.browserAction.setBadgeBackgroundColor({ color: "rgba(102,6,0,0.3)", tabId: tabId });
```
[the remark this answers is not in the document]
Comment (jonathan-s) [reply src/background.js]: This should be addressed in latest commit.

File: src/background.js (hunk 2 above)
Comment (sissbruecker) [reply src/background.js:30]: See discussion here: https://github.com/sissbruecker/linkding-extension/pull/47#discussion_r1278341611

IIRC it changes the permission model to be more granular (per domain), and Firefox shows a blue dot below the extension icon for each web page that the user has not given the extension permission for. To make things more confusing, the permission is automatically granted after clicking the extension icon, and the dot just disappears, leaving the user confused what the dot was even about. Then the dot shows up again for every other web page.

There is a way for the extension to request additional permissions for all domains after installation, but that just adds more friction and isn't really better than the v2 way of doing things. It also wasn't clear if requesting permissions for all domains was an intentional use of the API or not. 

Long story short, the way v3 works in Firefox makes the UX worse, at which point I decided to maintain the different implementations between Chrome and FF.

Comment (sissbruecker) [reply src/background.js:30]: Regarding the chrome branch, it only has a modified `manifest.json`, the remaining code in the main branch is made compatible with Manifest v3 by extracting browser-specific API into `browser.js`.

Comment (jonathan-s) [reply src/background.js:30]: Thank you, it sounds like for the time being keeping mv2 for Firefox is the way to go.

I found this forum post about the permissions api which you should use to listen if permissions has changed. https://discourse.mozilla.org/t/request-permissions-at-install-time-manifest-v3/121543/7

I assume that at some point mv2 will be deprecated so worth keeping an eye out. Though Mozilla hasn't yet set a timeline so a deprecation seems to be quite far into the future.

Comment (sissbruecker) [reply src/background.js]: Extracted APIs for interacting with the badge into `browser.js`
